enhancements/authentication/invalid-certs.md

```diff
@@ -0,0 +1,152 @@
+---
+title: invalid-certs
+authors:
+  - "@s-urbaniak"
+reviewers:
+  - "@sttts"
+  - "@stlaz"
+  - "@deads2k"
+approvers:
+  - "@mfojtik"
+creation-date: 2022-12-08
+last-updated: 2022-12-08
+---
+
+# Detecting invalid CA certificates
+
+## Summary
+
+OpenShift 4.10 is going to be rebased against Kubernetes 1.23. This requires using Go 1.17.
+However, starting with Go 1.17 support for legacy invalid is going to be removed, see https://go.dev/doc/go1.17.
+More formally, the temporary `GODEBUG=x509ignoreCN=0` flag has been removed.
+This implies that starting from OpenShift 4.10 invalid certificates will not be trusted any more
+as they will fail verification.
+
+This enhancement suggests a solution to prevent broken workloads from OpenShift 4.9 to OpenShift 4.10
+if invalid certificates are detected.
+
+This enhancment therefore proposes:
+- a way to detect invalid certificates in kube-apiserver and oauth-server
+- a way to prevent upgrades from OpenShift 4.9 to OpenShift 4.10 in the face of invalid certificates
```

Comment:
are we planning to backport an `Upgradeable=False` guard for this, or rely on the alert, or start with the alert and maybe follow up with `Upgradeable=False` if we see a lot of alerts, or...?

Reply:
@wking the suggestion is to backport an `Upgradeable=False` guard for this. An alert would be feasible if, after the upgrade, would be functional. However, in the presence of invalid certificates, a cluster will simply be dysfunctional after the upgrade to 4.10.

Reply:
> a way to prevent upgrades from OpenShift 4.9 to OpenShift 4.10 in the face of invalid certificates
For cases where the platform is a client, this is probably valid.  For cases where the platform is a server, but not a client, I think it's too strong to block upgrades. Can you separate the cases?

Reply:
> For cases where the platform is a client, this is probably valid. For cases where the platform is a server, but not a client, I think it's too strong to block upgrades. Can you separate the cases?

afaiu we have the following separation:

**platform is a client**
- custom API webhooks
- custom aggregated API endpoints
- certificates of external auth identity providers

**platform is a server**
- custom certificates for route endpoints
- custom serving certificates for kube-apiserver

However, if users configured an invalid custom serving cert for kube-apiserver in 4.9 and we don't block upgrades, communication to kube-apiserver will be broken in 4.10. We don't want to block upgrades in this case?

Reply:
i think all the cases enumerated here are platform clients:
- custom serving certificates for kube-apiserver: openshift workloads
- custom API webhooks: kube-apiserver being the client
- custom aggregated API endpoints: kube-apiserver being the client
- custom certificates for route endpoints: openshift oauth-proxy workloads being the client
- certificates of external auth identity providers: oauth-server being the client